### ISSUE TITLE
ptrace stability / usability in WSL? Esp. PTRACE_SETREGS / PTRACE_SYSCALL

### ISSUE THREAD

**Issue by ckahlo** [Microsoft/WSL#3978]
Hello there,

I commented on #1491 because I ran into the (basically) same bug. The ticket is rather old
(12/2016) and it didn't draw real attention so far, neither on the original issue nor on my
question. (no assignee, no project, no milestone)

* Your Windows build number:
1809 (Build 17763.437)

* What you're doing and what's happening:
Running a ptracer modifying the registers on PTRACE_SYSCALL (entering the syscall) and
updating with PTRACE_SETREGS. The register changes are not honoured. The _old_ syscall runs
with _old_ args.

* What's wrong / what should be happening instead:
Verified up to Linux kernel 4.15.0 I expect the register changes on syscall entry to be honoured
by the kernel. For reference the original implementation of the 4.4.0 kernel WSL claims to mimic:
https://elixir.bootlin.com/linux/v4.4/source/arch/x86/entry/entry_64.S#L237

Expectation: due to lack of functionality / buggy implementations in WSL I am intercepting and
modifying syscalls to get applications up and running - or at least track down the issues in detail. 
This works like a charm on original Linux kernels. I.e. introducing seccomp bpf with the help of a
ptracer. Sometimes dangerous (non-side-effect free) syscalls need to be skipped. To skip a
syscall orig_rax is changed to a known "not implemented" call like tuxcall (184), getpid, getuid, or
something like 0x00FFFFFF2342xxxx. The kernel responds with "ENOSYS" or something helpful,
but not dangerous and the ptracer takes care to deliver any actioncs needed on syscall exit. This
is not possible this way in WSL because WSL seems to rely on a 2nd stack copy of the register
set instead of the original register set. The changes made on syscall entry are visible in user
space after syscall return - but neither in the kernel nor on syscall exit.

* Strace of the failing command, if applicable:
Not applicable due to kernel mode.

* For WSL launch issues, please [collect detailed logs]
Not applicable, no launch issue, it's a runtime ptrace (debugging) issue.

**Comment by therealkenc** [comment]
> I commented on #1491 because I ran into the (basically) same bug. 

You're right. `PTRACE_SYSEMU` sure would be nice too.

**Comment by ckahlo** [comment]
PTRACE_SYSEMU is another story (not executing syscalls at all and not supported in mainline kernel).
Why is this issue closed? Is the behaviour of the emulated kernel fixed?

**Comment by therealkenc** [comment]
> duplicate to what?

The issue you cited. Even threw a 3rd like on there for ya.

> PTRACE_SYSEMU is ...  not supported in mainline kernel

`PTRACE_SYSEMU` has been in the [mainline kernel](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/kernel/ptrace.c#n1068) since circa 2005. But made me look anyway.

![image](https://user-images.githubusercontent.com/7049700/56270365-c954de80-60aa-11e9-8665-94fe9bb04113.png)


Whether you can get quality support service for a Debian Jessie userspace on a 5.1.0-rc5 UML running on top of Ubuntu's 4.18 kernel I couldn't tell ya. Depends on how much you were paying the guy, probably. 

But we digress. For the PTRACE_SYSCALL/PTRACE_SETREGS problem you want #1491.

**Comment by ckahlo** [comment]
Hi Ken,

thanks for clarification, didn't saw the like on the original post yet.
Quite interesting to know that PTRACE_SYSEMU is still available, I was (too) blindly following
http://manpages.ubuntu.com/manpages/bionic/en/man1/linux.uml.1.html (redirects to 14.04 LTS)
stating "To make it working, you need a kernel patch for your host, too.".
I see UML is available on Kali, but anyway UML is not the thing the customer needs.

Thank you for verifying and letting me know.

Best regards,
Christian

**Comment by therealkenc** [comment]
While you (we) wait on `ptrace(2)` improvements, maybe check out [pmem/syscall_intercept](https://github.com/pmem/syscall_intercept) which works for some (but not all) emulation scenarios.

> Sometimes dangerous (non-side-effect free) syscalls need to be skipped. 

Nah. Those are the ones worth shimming. 😉

**Comment by ckahlo** [comment]
Hi Ken,

had been AFK a bit. Thanks for the pointer. You're right, syscall_intercept is an option for some (emulation) scenarios. Had been there before because fakeroot uses a similar mechanism with LD pre-loading. Neither the less some of the executables are statically linked. And there's a need to protect the rest of the system against shellcode that uses syscalls directly. I was looking for seccomp bpf support at first. That's why I landed where I am, ptrace. ;-)

Best regards,
Christian
